Blogpost/stripe connect part 2 (#205)
ombulabs/blog: _posts/2019-10-11-stripe-connect-in-rails-part-2.md

---
layout: post
title: "Implementing Stripe Connect in Rails: Part 2"
date: 2019-10-11 14:00:00
categories: ["rails", "stripe"]
author: "luciano"
---

Some time ago we wrote [an article](https://www.ombulabs.com/blog/rails/stripe/stripe-connect-part-1.html) to explain how to implement [Stripe Connect](https://stripe.com/connect) in a [Rails](https://rubyonrails.org) application. That article covered mainly the connection part between Rails and Stripe. This one will cover the fun part, which is making transactions, charges, refunds, and more.

> If you haven't check the [first part](https://www.ombulabs.com/blog/rails/stripe/stripe-connect-part-1.html) yet I recommend you to do it since this article will continue from where that one ended.

Reviewer: ```suggestion
If you haven't checked out the [first part](https://www.ombulabs.com/blog/rails/stripe/stripe-connect-part-1.html) yet I recommend you do so since this article will continue from where that one ended.
```

<!--more-->

## Frontend

Before diving into the Stripe API, one of the first things that you should implement is a form that captures the payment information (e.g. credit card).
The simplest way to do this is using [Stripe Checkout](https://stripe.com/docs/payments/checkout), but if you want a more customizable version you can use [Stripe Elements](https://stripe.com/docs/web/setup).

> I recommend you to follow the official documentation of any these options so you have everything ready to continue with the backend work.

Reviewer: ```suggestion
I recommend that you follow the official documentation of any of these options so you have everything ready to continue with the backend work.
```

## Backend

### Charges

Once you have your form ready to collect payment information you'll probably want it to execute a [Charge](https://stripe.com/docs/api/charges) when it's submitted.

```ruby
# app/controllers/payments_controller.rb

class PaymentsController < ApplicationController
  # triggered after submit payment form
  def create
    # ...

    charge = StripeChargeService.charge!(params[:stripeToken])

    # ... (create Payment, store charge id in db, respond, etc.)
  end
end
```

> There are a few different [approaches](https://stripe.com/docs/connect/charges#choosing-approach) to implement the Stripe Charge. Each implementation fits better for different needs since in some cases the platform account should be the one that pays for the [Stripe fee](https://stripe.com/connect/pricing) and in some other cases it would be the connected account. I recommend you to take a look at the [official guides](https://stripe.com/docs/connect/charges) to make sure that you choose the one that fits better for you.

Reviewer: ```suggestion
There are a few different [approaches](https://stripe.com/docs/connect/charges#choosing-approach) to implement the Stripe Charge. Each implementation fits different needs, since in some cases the platform account should be the one that pays for the [Stripe fee](https://stripe.com/connect/pricing) and in some other cases it would be the connected account. I recommend that you take a look at the [official guides](https://stripe.com/docs/connect/charges) to make sure that you choose the one that fits better for you.
```

```ruby
# app/services/stripe_transfer_service.rb

class StripeTransferService
  def self.transfer!
    # ...

    Stripe::Transfer.create({
      amount: amount,
      currency: "usd",
      destination: user.stripe_user_id,
    })

    # ...
  end
end
```

To undo a Transfer that was previously made (either entirely or partially) you can use the [Transfer Reversal](https://stripe.com/docs/api/transfer_reversals) object.

If what you want is to transfer money the other way around (from a connected account to the platform account) you can use [Account Debits](https://stripe.com/docs/connect/account-debits), but keep in mind that there's an [additional cost](https://stripe.com/connect/pricing) for using it.

```ruby
Stripe::Transfer.create({
  amount: amount,
  currency: "usd",
  destination: "{PLATFORM_STRIPE_ACCOUNT_ID}",
  }, stripe_account: user.stripe_user_id)
```

### Refunds

With [Refunds](https://stripe.com/docs/api/refunds) you can undo any previous Charge that was made (either entirely or partially).

```ruby
# app/services/stripe_refund_service.rb

class StripeRefundService
  def self.refund!
    # ...

    Stripe::Refund.create({

>       charge: charge_token,

Reviewer 2: Maybe comment how to get this token?

Author: good idea 👍

      amount: amount,
      refund_application_fee: true
    }, stripe_account: user.stripe_user_id)

    # ...
  end
end
```

> By default it only refunds the portion of the Charge that went to the connected account, but you can set `refund_application_fee: true` to refund the platform account part too. Check out the [API documentation](https://stripe.com/docs/api/refunds) to see all the available attributes that you can use. Also keep in mind that the [Stripe fee](https://stripe.com/connect/pricing) that was paid by the platform account or connected account (depending of your configuration) is not refundable, so you're gonna lose that part if you perform a refund.

Reviewer: ```suggestion
By default it only refunds the portion of the Charge that went to the connected account, but you can set `refund_application_fee: true` to refund the platform account part too. Check out the [API documentation](https://stripe.com/docs/api/refunds) to see all the available attributes that you can use. Also keep in mind that the [Stripe fee](https://stripe.com/connect/pricing) that was paid by the platform account or connected account (depending on your configuration) is not refundable, so you will lose that part if you perform a refund.
```

### Conclusion

> This was an overview of some of the most popular Stripe object that are available. The Stripe API is very powerful and have a lot more things that you can do. So I recommend you to take a look to their guides for the things that were not covered in this article.

Reviewer: ```suggestion
This was an overview of some of the most popular Stripe objects that are available. The Stripe API is very powerful and has a lot more things that you can do. I recommend that you take a look at their guides for some things that were not covered in this article.
```
